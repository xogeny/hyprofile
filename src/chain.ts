import { JSONSchema4 } from "json-schema";

export class HyProfile {
    constructor(title?: string) {}
    resource(name: string): ResourceBuilder {
        return new ResourceBuilder(this);
    }
    action(name: string): ActionBuilder {
        return new ActionBuilder(this);
    }
    relation(name: string, ...details: Array<RelationDetails>): RelationBuilder {
        return new RelationBuilder(this);
    }
    schema(name: string, schema: JSONSchema4) {
        return new SchemaBuilder(this);
    }
}

export interface RelationDetails {
    from: string;
    to: string;
    min?: number;
    max?: number;
}

export class ResourceBuilder {
    constructor(protected profiler: HyProfile) {}
    resource(name: string): ResourceBuilder {
        return this.profiler.resource(name);
    }
    action(name: string): ActionBuilder {
        return this.profiler.action(name);
    }
    relation(name: string, ...details: Array<RelationDetails>): RelationBuilder {
        return this.profiler.relation(name);
    }
    schema(name: string, schema: JSONSchema4): SchemaBuilder {
        return this.profiler.schema(name, schema);
    }

    title(title: string): ResourceBuilder {
        return this;
    }
    description(title: string): ResourceBuilder {
        return this;
    }
    propertiesSchema(schema: JSONSchema4 | string): ResourceBuilder {
        return this;
    }
}

export class ActionBuilder {
    constructor(protected profiler: HyProfile) {}
    title(title: string): ActionBuilder {
        return this;
    }
    description(title: string): ActionBuilder {
        return this;
    }
    fields(schema: JSONSchema4 | string): ActionBuilder {
        return this;
    }
    method(method: "GET" | "PUT" | "POST" | "PATCH" | "OPTIONS" | "DELETE") {
        return this;
    }
    type(type: string) {
        return this;
    }
    action(name: string) {
        return this.profiler.action(name);
    }
    resource(name: string) {
        return this.profiler.resource(name);
    }
    relation(name: string, ...details: Array<RelationDetails>) {
        return this.profiler.relation(name);
    }
    schema(name: string, schema: JSONSchema4): SchemaBuilder {
        return this.profiler.schema(name, schema);
    }
}

export class RelationBuilder {
    constructor(protected profiler: HyProfile) {}
    relation(name: string, ...details: Array<RelationDetails>) {
        return this.profiler.relation(name);
    }
    schema(name: string, schema: JSONSchema4): SchemaBuilder {
        return this.profiler.schema(name, schema);
    }
}

export class SchemaBuilder {
    constructor(protected profiler: HyProfile) {}
    schema(name: string, schema: JSONSchema4): SchemaBuilder {
        return this.profiler.schema(name, schema);
    }
}

export const sample1 = new HyProfile("Order API")
    .resource("order")
    .title("Online Order")
    .description("An order submitted via the e-commerce API")
    .propertiesSchema({
        type: "object",
        properties: {
            orderNumber: {
                type: "number",
                description: "Order number used to cross reference with database and accounting",
            },
            itemCount: { type: "number" },
            status: { type: "string", enum: ["submitted", "pending", "completed"] },
        },
        required: ["orderNumber", "itemCount", "status"],
        additionalProperties: false,
    })
    .action("add-item")
    .title("Add Item")
    .method("POST")
    .type("application/x-www-form-urlencode")
    .fields("add-item")
    .resource("order-items")
    .resource("customer")
    .propertiesSchema("customer-properties")
    .relation("collection", { from: "order", to: "order-items" })
    .relation("items", { from: "order", to: "order-items" })
    .relation("customer", { from: "order", to: "customer" })
    .relation("info", { from: "order", to: "customer" })
    .relation("next", { from: "order", to: "order", max: 1 })
    .relation("prev", { from: "order", to: "order", max: 1 })
    .schema("add-item", {
        type: "object",
        properties: {
            orderNumber: { type: "number" },
            productCode: { type: "string" },
            quantity: { type: "number" },
        },
        additionalProperties: false,
    })
    .schema("customer-properties", {
        type: "object",
        properties: {
            customerId: { type: "string" },
            name: { type: "string" },
        },
        required: ["customerId", "name"],
        additionalProperties: false,
    });
