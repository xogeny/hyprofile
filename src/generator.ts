import fs from "fs";
import path from "path";

import debug from "debug";
const debugWriter = debug("hyprofile:writer");

export interface Generator {
    writeFile(filename: string, contents: string): Promise<void>;
}

export class FileWriter implements Generator {
    constructor(public directory: string) {}
    writeFile(filename: string, contents: string): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            let file = path.join(this.directory, filename);
            fs.writeFile(file, contents, err => {
                if (err) {
                    console.error(`Error while writing ${file}: ${err}`);
                    reject(err);
                } else {
                    debugWriter("File %s written", file);
                    debugWriter("  Contents: %s", contents);
                    resolve(undefined);
                }
            });
        });
    }
}

export class MemoryWriter implements Generator {
    files: { [name: string]: string } = {};
    writer: Generator | undefined = undefined;
    constructor(directory?: string) {
        if (directory) {
            this.writer = new FileWriter(directory);
        }
    }
    async writeFile(filename: string, contents: string): Promise<void> {
        this.files[filename] = contents;
        if (this.writer) await this.writer.writeFile(filename, contents);
        return undefined;
    }
}
