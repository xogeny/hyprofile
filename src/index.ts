export * from "./profile";
export * from "./sample";
export * from "./types";
export * from "./documentation";
