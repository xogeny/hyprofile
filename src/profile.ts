import { ISchema } from "@xogeny/ts-schema";

export type SchemaReference = { $ref: string };

export type Schema = ISchema | SchemaReference;

export interface Profile {
    resources?: { [className: string]: ClassProfile };
    relations?: { [rel: string]: Array<RelationProfile> };
    schemas?: { [schema: string]: Schema };
}

export interface ClassProfile {
    properties?: Schema;
    actions?: { [actionName: string]: ActionProfile };
}

export interface ActionProfile {
    title?: string;
    method?: "GET" | "POST" | "PUT" | "PATCH" | "OPTIONS" | "DELETE";
    type?: string;
    fields?: Schema;
}

export interface RelationProfile {
    from: string;
    to: string[];
}
