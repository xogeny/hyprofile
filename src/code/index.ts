import { ElaboratedProfile } from "../elaborate";
import { Generator } from "../generator";

export async function generateCode(profile: ElaboratedProfile, generator: Generator) {
    let contents = "";
    profile.resources.forEach(resource => {
        contents = `${contents}\n${resource.typedef}`;
    });
    await generator.writeFile("types.ts", contents);
}
