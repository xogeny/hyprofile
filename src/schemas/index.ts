import { Generator } from "../generator";
import { ElaboratedProfile } from "../elaborate";
import { JSONSchema4 } from "json-schema";

interface ResourceData {
    propSchema: JSONSchema4;
    actionSchemas: { [name: string]: JSONSchema4 };
}

interface SchemaData {
    resources: { [id: string]: ResourceData };
}

export async function generateSchemas(profile: ElaboratedProfile, generator: Generator): Promise<void> {
    let data: SchemaData = { resources: {} };
    profile.resources.forEach(resource => {
        data.resources[resource.resource] = {
            propSchema: resource.schema,
            actionSchemas: {},
        };
    });
    await generator.writeFile("schemas.json", JSON.stringify(data, null, 4));

    return undefined;
}
