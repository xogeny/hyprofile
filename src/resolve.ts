import { JSONSchema4 } from "json-schema";
import { dereference } from "json-schema-to-typescript/dist/src/resolver";
import { Schema, Profile } from "./profile";

export function resolveSchema(schema: Schema, profile: Profile): Promise<JSONSchema4> {
    const cwd = process.cwd() + "/";
    console.log("Current directory: ", cwd);
    if (typeof schema === "string") {
        let id = schema;
        if (profile.schemas) {
            schema = profile.schemas[id];
            return dereference(schema, cwd);
        } else {
            throw new Error("No local schema for " + id);
        }
    }
    return dereference(schema, cwd);
}
