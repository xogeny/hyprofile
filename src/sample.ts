import { Profile } from "./profile";

export namespace Samples {
    export const sample1: Profile = {
        resources: {
            order: {
                properties: {
                    type: "object",
                    properties: {
                        orderNumber: { type: "number" },
                        itemCount: { type: "number" },
                        status: { type: "string", enum: ["submitted", "pending", "completed"] },
                    },
                    required: ["orderNumber", "itemCount", "status"],
                },
                actions: {
                    "add-item": {
                        title: "Add Item",
                        method: "POST",
                        type: "application/x-www-form-urlencoded",
                        fields: { $ref: "#/schemas/add-item" },
                    },
                },
            },
            "order-items": {},
            customer: {
                properties: { $ref: "#/schemas/customer-properties" },
            },
        },
        relations: {
            collection: [{ from: "order", to: ["order-items"] }],
            items: [{ from: "order", to: ["order-items"] }],
            customer: [{ from: "order", to: ["customer"] }],
            info: [{ from: "order", to: ["customer"] }],
            next: [{ from: "order", to: ["order"] }],
            prev: [{ from: "order", to: ["order"] }],
        },
        schemas: {
            "add-item": {
                type: "object",
                properties: {
                    orderNumber: { type: "number" },
                    productCode: { type: "text" },
                    quantity: { type: "number" },
                },
            },
            "customer-properties": {
                type: "object",
                properties: {
                    customerId: { type: "string" },
                    name: { type: "string" },
                },
                required: ["customerId", "name"],
            },
        },
    };
}
