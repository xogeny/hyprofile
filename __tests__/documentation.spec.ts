import * as hyprofile from "../src";

describe("Test Documentation Generation", () => {
    it("should generate documentation for sample profile", async () => {
        let profile = hyprofile.Samples.sample1;
        let result = await hyprofile.generateDocumentation(profile);
        expect(result).toMatchSnapshot();
    });
});
