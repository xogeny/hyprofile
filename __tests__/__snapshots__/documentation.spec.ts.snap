// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Documentation Generation should generate documentation for sample profile 1`] = `
"<div><h1>Documentation</h1><h2>Resources</h2><div><h3><code>order</code></h3><pre>/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

export interface Order {
  orderNumber: number;
  itemCount: number;
  status: &quot;submitted&quot; | &quot;pending&quot; | &quot;completed&quot;;
  [k: string]: any;
}
</pre></div><div><h3><code>customer</code></h3><pre>/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

export interface Customer {
  customerId: string;
  name: string;
  [k: string]: any;
}
</pre></div></div>"
`;
